src/models/TeamGenerator.ts

```diff
@@ -314,43 +353,54 @@ export default class BalancedRandomTeamGenerator implements TeamGenerator {
 
   /**
    * This function takes a list of teams and optimizes the balancing between them.
-   * This is done by sorting the teams from best to worst and iteratively try to
-   * swap 1 optimum player pair of the best and worst team and reshuffle the team
-   * array.
+   * This is done by iteratively try to swap 1 optimum player pair between any team combination
+   * until there's no more improvment possible.
    *
    * The single player swap is based on a greedy approach (local optimum) but since
    * only one optimum player is swapped and not the whole team composition between
-   * best and worst team this optimization algorithm may aim for a global balance
-   * optimum over all teams (to be verified).
+   * any two teams this aims for a more global optimum. Especially since now any team
+   * combination is compared instead of best and worst.
    *
    * Again this function assumes validation has been done to assure that all players
    * within the team have their game skill assessed for the given game!!!!
    *
    * @param teams The teams for which to optimize their balance
    */
   protected optimizeTeamSkillBalance(teams: Array<Team>): void {
-    const teamAscendingComparer = (team1: Team, team2: Team): number => {
-      return team1.getTeamGameSkill() - team2.getTeamGameSkill();
-    };
+    if (teams.length < 2){
+      return;
+    }
 
-    const teamSize: number = teams[0].targetSize; // teams param must not be empty!!
-    let maxSwapsLeft: number = teams.length * teams.length * (teamSize - 1); // (team size - 1) swaps per team pair.
     let lastSwapSuccessful: boolean = true;
 
-    // optimize team balance as long as successful but at most until all swap options are exhausted
-    while (lastSwapSuccessful && maxSwapsLeft > 0) {
-      teams.sort(teamAscendingComparer);
-      // try to swap one player between best and worst team:
-      lastSwapSuccessful = this.trySwapPlayerForBetterBalance(
-        teams[0],
-        teams[teams.length - 1],
-      );
-      maxSwapsLeft--;
+    // optimize team balance as long as successful. Will eventually stop as no swapping improvement possible anymore.
+    while (lastSwapSuccessful) {
+
+      lastSwapSuccessful = false; // safeguard to terminate while loop.
+      for (let teamA of teams){
+        for (let teamB of teams){
+          if (teamA !== teamB){// try to swap one player between two different teams
+            lastSwapSuccessful = this.trySwapPlayerForBetterBalance(
+              teamA,
+              teamB,
+            );
+          }
+          if (lastSwapSuccessful){break;} // break for outer while loop on success. 
+          // if not breaking then possible optimization could be overseen when exiting for loops with last attempt == false.
+        }
+        if (lastSwapSuccessful){break;} // break for outer while loop on success.
+      }
     }
   }
 
+  protected filterBySkill(teams: Array<Team>, skill: number): Array<Team> {
+
```

Comment:
Unnecessary line breaks

Reply:
Strange, I did not touch this function at all. Anyway I will remove it as 
1) it is not used
2) does not do anything.